### DISCUSSION TITLE
It's that useful if we create a TypedAppender?

### DISCUSSION THREAD

**Discussion by fililili** [duckdb/duckdb#12100] [Ideas]
When I read code, I found append function will check type like below:
```cpp
template <>
void BaseAppender::Append(bool value) {
	AppendValueInternal<bool>(value);
}
```
```cpp
void BaseAppender::AppendValueInternal(T input) {
	if (column >= types.size()) {
		throw InvalidInputException("Too many appends for chunk!");
	}
	auto &col = chunk.data[column];
	switch (col.GetType().id()) {
	case LogicalTypeId::BOOLEAN:
		AppendValueInternal<T, bool>(col, input);
		break;
	case LogicalTypeId::UTINYINT:
		AppendValueInternal<T, uint8_t>(col, input);
		break;
```
The type check will check what `col.GetType().id()` is, if `col.GetType().id()` is `LogicalTypeId::BOOLEAN`, then it call `AppendValueInternal<T, bool>(col, input);`. And such type check will happen every time we call append function, it should be time consuming.
However, can we add a template class TypedAppender<Types...>, for example, suppose we have `TypeAppender<bool, uint8_t>` , then we can provide an appenderRow function in this way:
```cpp
void TypeAppender::appenderRow(T1 val1, T2 val2) {
	BeginRow();
	AppendValueInternal<T1, bool>(0, val1);
	AppendValueInternal<T1, uint8_t>(0, val2);
	EndRow();
}
```
so we don't have to do type check every time we call append function, performance should be better. (btw, only c++ programmer can benefit from it.)
The only risk is we may write wrong types for a table. But it's not a really problem, we can add some check in the constructor of TypeAppender.

**Comment by Tishj** [comment]
You'll have to be more specific, where is this type check taking place?

**Reply by fililili** [reply]
have updated some more details, thanks

**Reply by Tishj** [reply]
I think we have AppendDataChunk for that reason
